### DISCUSSION TITLE
Is there a keyboard shortcut to switch panes

### DISCUSSION THREAD

**Discussion by aphsa** [wezterm/wezterm#3331] [Q&A] [answered]
Noob here. Is there a keyboard shortcut to switch panes?
"wezterm show-keys" doesn't have one.

**Comment by wez** [comment, marked as answer]
What do you mean by switch? 

There are actions for activating panes:
* https://wezfurlong.org/wezterm/config/lua/keyassignment/ActivatePaneDirection.html
* https://wezfurlong.org/wezterm/config/lua/keyassignment/ActivatePaneByIndex.html

You can also rotate panes:
* https://wezfurlong.org/wezterm/config/lua/keyassignment/RotatePanes.html

You can use PaneSelect mode to activate and swap panes:
* https://wezfurlong.org/wezterm/config/lua/keyassignment/PaneSelect.html

**Reply by simonmandlik** [reply]
@wez I think OP is referring to an action similar to e.g. `swap-pane -{D, U}` in tmux, which switches active pane with the one below/above it, keeping the focus.

My attempt to achieve something similar is here:

```lua
local function switch_in_direction(dir)
    return function(window)
        local tab = window:active_tab()
        local next_pane = tab:get_pane_direction(dir)
        if next_pane then
            tab.swap_active_with_index(next_pane, true)
        end
    end
end
```

and then

```lua
config.keys = {
    { mods = "LEADER", key = "J",
        action = wezterm.action_callback(switch_in_direction("Down"))
    },
    { mods = "LEADER", key = "K",
        action = wezterm.action_callback(switch_in_direction("Up"))
    }
}

but `MuxTab` returned by `active_tab()` doesn't have `swap_active_with_index`.

**Reply by xiangpeng2020** [reply]
agree with @simonmandlik

**Comment by whalelogic** [comment]
I think he is asking how to change the active window. You want to use Ctrl+Shift+Arrow(left/right/up/down).

**Comment by akharrou** [comment]
can you please expose something like `swap_active_with_index` and `swap_active_with_index_keep_focus` so we can get swap functionality like tmux or kitty?

this works, just missing such an api

```lua
local pane_select_alphabet = '123456789'

local function swap_pane_direction(direction)
  return wezterm.action_callback(function(window, pane)
    local tab = pane:tab()
    if not tab then
      return
    end

    local target = tab:get_pane_direction(direction)
    if not target then
      return
    end

    local target_id = target:pane_id()
    local label = nil

    for index, pane_info in ipairs(tab:panes_with_info()) do
      if pane_info.pane:pane_id() == target_id then
        label = pane_select_alphabet:sub(index, index)
        break
      end
    end

    if not label or label == '' then
      return
    end

    window:perform_action(act.PaneSelect {
      mode = 'SwapWithActiveKeepFocus',
      alphabet = pane_select_alphabet,
    }, pane)

    wezterm.time.call_after(0.05, function()
      window:perform_action(act.SendKey { key = label }, pane)
    end)
  end)
end
```
